src/results/marginalization.rs

```diff
@@ -73,3 +81,140 @@ pub fn marginal_distribution(
+    memory: Vec<String>,
+    indices: Option<Vec<usize>>,
+    return_int: bool,
+    return_hex: bool,
+    parallel_threshold: usize,
+) -> PyResult<PyObject> {
+    let run_in_parallel = getenv_use_multiple_threads();
+    let first_elem = memory.get(0);
+    if first_elem.is_none() {
+        let res: Vec<String> = Vec::new();
+        return Ok(res.to_object(py));
+    }
+
+    let clbit_size = hex_to_bin(first_elem.unwrap()).len();
+
+    let out_mem: Vec<String> = if memory.len() < parallel_threshold || !run_in_parallel {
+        memory
+            .iter()
+            .map(|x| map_memory(x, &indices, clbit_size, return_hex))
+            .collect()
+    } else {
+        memory
+            .par_iter()
+            .map(|x| map_memory(x, &indices, clbit_size, return_hex))
+            .collect()
+    };
+    if return_int {
+        if out_mem.len() < parallel_threshold || !run_in_parallel {
+            Ok(out_mem
+                .iter()
+                .map(|x| BigUint::parse_bytes(x.as_bytes(), 2).unwrap())
+                .collect::<Vec<BigUint>>()
+                .to_object(py))
+        } else {
+            Ok(out_mem
+                .par_iter()
+                .map(|x| BigUint::parse_bytes(x.as_bytes(), 2).unwrap())
+                .collect::<Vec<BigUint>>()
+                .to_object(py))
+        }
+    } else {
+        Ok(out_mem.to_object(py))
+    }
+}
+
+#[pyfunction]
+pub fn marginal_measure_level_0(
+    py: Python,
+    memory: PyReadonlyArray3<Complex64>,
+    indices: Vec<usize>,
+) -> PyObject {
+    let mem_arr: ArrayView3<Complex64> = memory.as_array();
+    let input_shape = mem_arr.shape();
+    let new_shape = [input_shape[0], indices.len(), input_shape[2]];
+    let out_arr: Array3<Complex64> =
+        Array3::from_shape_fn(new_shape, |(i, j, k)| mem_arr[[i, indices[j], k]]);
+    out_arr.into_pyarray(py).into()
+}
+
+#[pyfunction]
+pub fn marginal_measure_level_0_avg(
+    py: Python,
+    memory: PyReadonlyArray2<Complex64>,
+    indices: Vec<usize>,
+) -> PyObject {
+    let mem_arr: ArrayView2<Complex64> = memory.as_array();
+    let input_shape = mem_arr.shape();
+    let new_shape = [indices.len(), input_shape[1]];
+    let out_arr: Array2<Complex64> =
+        Array2::from_shape_fn(new_shape, |(i, j)| mem_arr[[indices[i], j]]);
+    out_arr.into_pyarray(py).into()
+}
+
+#[pyfunction]
+pub fn marginal_measure_level_1(
+    py: Python,
+    memory: PyReadonlyArray2<Complex64>,
+    indices: Vec<usize>,
+) -> PyObject {
+    let mem_arr: ArrayView2<Complex64> = memory.as_array();
+    let input_shape = mem_arr.shape();
+    let new_shape = [input_shape[0], indices.len()];
+    let out_arr: Array2<Complex64> =
+        Array2::from_shape_fn(new_shape, |(i, j)| mem_arr[[i, indices[j]]]);
+    out_arr.into_pyarray(py).into()
+}
+
+#[pyfunction]
+pub fn marginal_measure_level_1_avg(
+    py: Python,
+    memory: PyReadonlyArray1<Complex64>,
+    indices: Vec<usize>,
+) -> PyResult<PyObject> {
```

Comment:
Other functions don't use `PyResult` return type. Is this intended?

Reply:
It is intended it's a side effect of L217, level 1 avg data is the only input that's a 1d array. The `as_slice()` method on the read only numpy array view returns a result if it can't build a slice from the numpy array (basically if it's non contiguous I think) so we need to return that to python if `as_slice()` returns an error.